Brooklyn OSGification - step 1 (brooklyn-core)

CMoH opened the pull request:
Create base projects for karaf, next work on karaf features until brooklyn-core is installable in karaf.

I've created the PR for review, leaving all commits as they were so one can understand the intention of each change. However, I would like to reword them and reference an OSGification jira issue.

hzbarcea commented:
@CMoH , I don't know why Jenkins failed, from a cursory look at the output logs it shouldn't have happened. Something to investigate.

Tests look ok and a quick look at the changes also looks ok. I also tested with a clean .m2 repository with no problems. I will spend a good part of tomorrow reviewing carefully your changes and test for regression issues. This is a big change and we should be careful with merging. I would think that we should try to merge this sooner and continue testing and fix regression issues until 0.9.0, but I am curious what others think too. Extra pairs of eyes would be extremely helpful.

splatch commented:
:+1: Looks good for me. Builds and runs. There are small improvements which can be added later on to make things nicer but lets merge this for now. Ie. manifest generation can be improved via maven plugin management, wrapped files could be also repackaged (if it's worth ofc), and features file can be filtered and there is existing reflections bundle from apache servicemix which should be already released. But all these things are really minor. Go for it. :-)

hzbarcea commented:
I ran more tests, didn't see any regression issue, but it's not impossible for something to have crept in, something to keep an eye on in the coming days. I am merging this and let's make the suggested improvements in subsequent PRs. @splatch , thanks, expecting some PRs from you :).

neykov commented:
@hzbarcea, @CMoH The master build is failing after merging this PR, could you take a look.

ahgittin commented on core/src/main/java/org/apache/brooklyn/core/mgmt/persist/XmlMementoSerializer.java:
```diff
@@ -345,7 +345,36 @@ public Object unmarshal(HierarchicalStreamReader reader, UnmarshallingContext co
+    // Perhaps context.getRequiredType(); can be used instead?
+    // Other users of xstream (e.g. jenkinsci) manually check for resoved-to and class attributes
+    //   for compatibility with older versions of xstream
+    public static Class readClassType(HierarchicalStreamReader reader, Mapper mapper) {
```
should this and next method be `private static`

ahgittin commented on core/src/main/java/org/apache/brooklyn/util/core/osgi/Osgis.java:
```diff
@@ -90,46 +69,8 @@
-    private static final String MANIFEST_PATH = "META-INF/MANIFEST.MF";
     private static final Set<String> SYSTEM_BUNDLES = MutableSet.of();
 
-    public static class VersionedName {
```
violates deprecation policy -- mark as `/** @deprecated since 0.9.0, replaced with top-level class */`.  this impl could extend the top-level class.  that will ensure code using this does not break.

ahgittin commented on core/src/main/java/org/apache/brooklyn/util/core/osgi/Osgis.java:
```diff
@@ -336,195 +277,56 @@ public static BundleFinder bundleFinder(Framework framework) {
+     * @todo Use felixCacheDir ?
      */
-    
-    public static FrameworkFactory newFrameworkFactory() {
```
keep for compatibility, but deprecated

ahgittin commented on core/src/main/java/org/apache/brooklyn/util/core/osgi/Osgis.java:
```diff
@@ -336,195 +277,56 @@ public static BundleFinder bundleFinder(Framework framework) {
-        throw new IllegalStateException("Could not find framework factory.");
-    }
-    
-    public static Framework newFrameworkStarted(String felixCacheDir, boolean clean, Map<?,?> extraStartupConfig) {
```
keep for compatibility, but deprecated

ahgittin commented on core/src/main/java/org/apache/brooklyn/util/core/osgi/Osgis.java:
```diff
@@ -336,195 +277,56 @@ public static BundleFinder bundleFinder(Framework framework) {
-                try {in.close();} 
-                catch (Exception e) {};
-            }
+    public static void ungetFramework(Framework framework) {
```
is `unget` a technical term?  deserves javadoc, or make it private.

ahgittin replied:
nvm - i see the javadoc above.  is `stop` a better name than `unget`?  or will frameworkers know the term?

CMoH replied:
The lifetime of the OSGi framework will not be controlled by Brooklyn, so it makes sense to obtain the framework similarly to other OSGi services. Those methods are get/ungetService, so I derived the names get/ungetFramework. However, when running within the brooklyn launcher their effect is start/stop, maintaining the previous semantics.

ahgittin commented on core/src/main/java/org/apache/brooklyn/util/core/osgi/Osgis.java:
```diff
@@ -623,97 +425,4 @@ public static boolean isExtensionBundle(Bundle bundle) {
-    /** Takes a string which might be of the form "symbolic-name" or "symbolic-name:version" (or something else entirely)
-     * and returns a VersionedName. The versionedName.getVersion() will be null if if there was no version in the input
-     * (or returning {@link Maybe#absent()} if not valid, with a suitable error message). */
-    public static Maybe<VersionedName> parseOsgiIdentifier(String symbolicNameOptionalWithVersion) {
```
keep for compatibility, but deprecated

ahgittin commented on core/src/main/java/org/apache/brooklyn/util/core/osgi/Osgis.java:
```diff
@@ -623,97 +425,4 @@ public static boolean isExtensionBundle(Bundle bundle) {
-     * The class is not used, staying for future reference.
-     * Remove after OSGi transition is completed.
-     */
-    public static class ManifestHelper {
```
keep for compatibility, but deprecated

ahgittin commented on karaf/apache-brooklyn/src/main/resources/etc/org.ops4j.pax.logging.cfg:
```diff
@@ -0,0 +1,46 @@
+################################################################################
```
do we need this file?  we already have a sophisticated `log4j.xml` config hierarchy as per website.  what is relationship of this with that?  if we keep this it needs to be explained, and maybe docs updated.

ahgittin commented on karaf/commands/pom.xml:
```diff
@@ -0,0 +1,77 @@
+  </parent>
+
+    <groupId>org.apache.brooklyn</groupId>
+    <artifactId>brooklyn-commands</artifactId>
```
`brooklyn-karaf-commands` ?

hzbarcea replied:
This is a placeholder project for [karaf commands](https://karaf.apache.org/manual/latest/commands/commands.html) one could use in the karaf console. Yes, it should be `@Beta`.

ahgittin commented on karaf/commands/src/main/java/org/apache/brooklyn/karaf/commands/Catalog.java:
```diff
@@ -0,0 +1,44 @@
+         System.out.println("Executing command catalog");
+         System.out.println("Option: " + option);
+         System.out.println("Argument: " + argument);
+         return null;
```
i guess this is experimental?  can you label the class as `@Beta` ?

ahgittin commented on karaf/features/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,89 @@
+    See the License for the specific language governing permissions and
+    limitations under the License.
+-->
+<features xmlns="http://karaf.apache.org/xmlns/features/v1.2.0" name="org.apache.brooklyn-${project.version}">
```
yikes, do we need to maintain this?  can't it be auto-generated from hints in the pom?  feels very brittle.

if it does need maintenance can this have instructions as a comment, at least enough to point people at the right resources and intention.

CMoH replied:
This file contains the runtime bundles dependencies for each karaf "feature". As far as I know there is work being done to automating the process from maven dependencies, but I'm not sure.

For now I am focusing on creating a workable monolithic karaf feature that encompasses brooklyn-core, the rest server and jsgui project. However, I expect this file to become more and more slim and easier to read as we advance down the OSGi path.

ahgittin commented on karaf/features/src/main/history/dependencies.xml:
```diff
@@ -0,0 +1,47 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<features xmlns="http://karaf.apache.org/xmlns/features/v1.3.0" name="org.apache.brooklyn-0.9.0-SNAPSHOT">
+    <feature version="0.0.0">
```
and for this -- do we need to maintain it?  also, why is it in `history` ?

CMoH replied:
This is generated by karaf-maven-plugin, and it is generated in src rather than target for the purpose of controlling changes within runtime dependencies. Arguments within pom.xml can get rid of it if we find it is not useful.

ahgittin commented on karaf/itest/src/test/java/org/apache/brooklyn/AssemblyTest.java:
```diff
@@ -0,0 +1,117 @@
+
+import org.apache.karaf.features.BootFinished;
+import org.apache.karaf.features.FeaturesService;
+import org.junit.Test;
```
we use `testng` everywhere else.  can we avoid an add'l `junit` dependency?

CMoH replied:
I've modified the test to use testng in #992. I'm not sure it's the right change; `pax-exam` has better support for `junit`, but it can be an opportunity to enhance its `testng` support.

ahgittin commented on pom.xml:
```diff
@@ -133,6 +245,10 @@
         <module>utils/test-support</module>
         <module>utils/rest-swagger</module>
+
+        <module>utils/rt-osgi</module>
```
what is `rt`?  can we have a clearer project name?

CMoH replied:
`rt` is shorthand for `runtime`, but maybe utils might be a better name?

ahgittin commented on utils/rt-osgi/src/test/java/org/apache/brooklyn/util/osgi/OsgisTest.java:
```diff
@@ -30,12 +28,12 @@
-        assertFalse(Osgis.parseOsgiIdentifier("a.b:0.1.2:3.4.5").isPresent());    // too many colons
-        assertFalse(Osgis.parseOsgiIdentifier("a.b:0.0.0_SNAPSHOT").isPresent()); // invalid version
-        assertFalse(Osgis.parseOsgiIdentifier("").isPresent());
+        assertEquals(OsgiUtils.parseOsgiIdentifier("a.b").get(), new VersionedName("a.b", null));
```
so now we have both `Osgis` and `OsgiUtils`?  that's confusing.  let's just have one, or give them more detailed names.  (but remember we want to preserve backwards compatibility.)

CMoH replied:
The rt-osgi project is a temporary repository for code that is needed both by brooklyn-core and rt-felix. `OsgiUtils` is the class where code was moved from `Osgis`; the name change is supposed to make it easier to distinguish which code needs which project.

The plan is to move it back to core when we decide to drop rt-felix, but... I guess we'll see when we get there.

ahgittin commented:
great set of changes to add karaf!

lots of quite minor comments but ideally they should have been addressed before merging.  we don't like to make master backwards incompatible without a deprecation cycle. and in this case i think it could have been avoided.

@CMoH is it going to be possible to address these comments *and* fix the issue @neykov refers to today?  if not i suggest we revert until they can be addressed.

ahgittin commented:
BTW it builds locally fine for me, but on our Apache Builder server it fails, e.g. https://builds.apache.org/job/incubator-brooklyn-master-build/616/console .  The cause looks to be:

```
INFO] --- karaf-maven-plugin:4.0.1:assembly (default-assembly) @ apache-brooklyn ---
[INFO] Creating work directory
[INFO] Loading kar and features repositories dependencies
[INFO] Generating karaf assembly: /jenkins/workspace/incubator-brooklyn-master-build/karaf/apache-brooklyn/target/assembly
[INFO] Unzipping kars
[WARNING] Error resolving artifactorg.apache.brooklyn:brooklyn-features:xml:features:0.9.0-SNAPSHOT:Could not find artifact org.apache.brooklyn:brooklyn-features:xml:features:0.9.0-SNAPSHOT
shaded.org.eclipse.aether.resolution.ArtifactResolutionException: Could not find artifact org.apache.brooklyn:brooklyn-features:xml:features:0.9.0-SNAPSHOT
	at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:444)
	at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:246)
	at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifact(DefaultArtifactResolver.java:223)
	at shaded.org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveArtifact(DefaultRepositorySystem.java:294)
	at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:573)
	at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:528)
	at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:506)
	at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:481)
	at org.apache.karaf.features.internal.download.impl.MavenDownloadTask.download(MavenDownloadTask.java:34)
	at org.apache.karaf.features.internal.download.impl.AbstractRetryableDownloadTask.run(AbstractRetryableDownloadTask.java:58)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: shaded.org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact org.apache.brooklyn:brooklyn-features:xml:features:0.9.0-SNAPSHOT
	at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:434)
	... 16 more
```

hzbarcea commented:
@neykov , @ahgittin a few of us tested a number of times and I reviewed it with @splatch yesterday.

The build failure does not make much sense, as the features project is built right before, in the Jenkin's log. Do you get the same error or another? Btw, I also tested myself with a completely clean .m2 repository. That says, we obviously need to identify and fix the issue.

neykov commented:
@hzbarcea Using the exact same maven command line arguments might help in reproducing this, you can find them in the console output of the build. Also Java version might influence (unlikely) this - the apache builds use Java 7.

CMoH commented:
I am using Java7 on my machine, so I don't think this can be the reason.

splatch commented:
Artifact resolution is made via maven repository. From logs I see that Jenkins setup it's other than default with ~/.m2/repository, thus I suspect that as the trouble maker.

CMoH commented:
I'm trying to build right now on my machine using the same maven arguments as I have seen in jenkins' console output. So far maven-3.3.3 built the project successfully. maven-3.0.4 failed the first build at the brooklyn-policy project, and a second run (again with a clean repo) produced a successful build (including the karaf assembly, everything). Maybe it's a non-deterministic failure? Has 

Here is my test command: `mvn -B -f /home/cipi/src/incubator-brooklyn/pom.xml -Dmaven.repo.local=/home/cipi/.m2tmp/0 --also-make clean install`

As far as I can see, the error can occur from two places, which can be seen in the log:

1. brooklyn-features generates the features.xml using karaf-maven-plugin and install it as an artifact in the maven repo as seen here in the log:

```
[INFO] --- maven-install-plugin:2.5.2:install (default-install) @ brooklyn-features ---
[INFO] No primary artifact to install, installing attached artifacts instead.
[INFO] Installing /jenkins/workspace/incubator-brooklyn-master-build/karaf/features/pom.xml to /jenkins/maven-repositories/0/org/apache/brooklyn/brooklyn-features/0.9.0-SNAPSHOT/brooklyn-features-0.9.0-SNAPSHOT.pom
[INFO] Installing /jenkins/workspace/incubator-brooklyn-master-build/karaf/features/target/feature/feature.xml to /jenkins/maven-repositories/0/org/apache/brooklyn/brooklyn-features/0.9.0-SNAPSHOT/brooklyn-features-0.9.0-SNAPSHOT-features.xml
```

2. The karaf assembly project "apache-brooklyn" tries to locate the above artifact through karaf-maven-plugin, but fails to retrieve it from the maven repository:

```
[WARNING] Error resolving artifactorg.apache.brooklyn:brooklyn-features:xml:features:0.9.0-SNAPSHOT:Could not find artifact org.apache.brooklyn:brooklyn-features:xml:features:0.9.0-SNAPSHOT
shaded.org.eclipse.aether.resolution.ArtifactResolutionException: Could not find artifact org.apache.brooklyn:brooklyn-features:xml:features:0.9.0-SNAPSHOT
	at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:444)
........
Caused by: shaded.org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact org.apache.brooklyn:brooklyn-features:xml:features:0.9.0-SNAPSHOT
```

I cannot spot any logical errors at this level, but I would like to look at the files on the build machine and see if features.xml is actually installed in the repository, so we can determine if the error occurs at step 1 or step 2. Can someone with access rights verify this? If it is there, we need to determine why the installed artifact is not retrievable (NFS mount? plugin bug? plugin bug on maven-3.0.4?)

@ahgittin I'll start addressing your comments in the meantime.

hzbarcea commented:
@cmoh, I think what @splatch is saying is that while mvn has been told to use a different maven repo via the -Dmaven.repo.local param, karaf hasn't been informed of the same while running tests (with paxexam) and therefore will try to resolve artifacts from the default local maven repo.

Let me try to find the incantation for that. Hopefully @splatch knows it by heart and will beat me to it.

CMoH commented:
@hzbarcea @splatch  That is quite insightful. However, this might mean we actually have 2 problems, since the error within jenkins occurs during build phase, not test phase. Could it be that karaf-maven-plugin does not honor -Dmaven.repo.local ? I'll test right away

splatch commented:
@hzbarcea @CMoH Yeah, if there is custom repository location pax url (and pax exam running karaf) will get crazy, but what now fails is building a distro. Karaf maven plugin shoud honor maven.repo.local property.

neykov commented:
@hzbarcea The build is on apache infra, so don't think anyone can access the machine. But looks like you've got a lead and that's not needed any more.

CMoH commented:
The build took quite a while now, but I have results after running the same maven build as above while having the default repository empty. Unfortunately some of the plugins do not honor -Dmaven.repo.local, producing the following files in the default mvn repo:

```
cipi@pink ~/.m2 $ find . -type d
.
./repository
./repository/org
./repository/org/apache
./repository/org/apache/karaf
./repository/org/apache/karaf/features
./repository/org/apache/karaf/features/enterprise
./repository/org/apache/karaf/features/enterprise/4.0.1
./repository/org/apache/karaf/features/standard
./repository/org/apache/karaf/features/standard/4.0.1
./repository/org/apache/karaf/features/framework
./repository/org/apache/karaf/features/framework/4.0.1
./repository/org/ops4j
./repository/org/ops4j/pax
./repository/org/ops4j/pax/web
./repository/org/ops4j/pax/web/pax-web-features
./repository/org/ops4j/pax/web/pax-web-features/4.2.0
./repository/org/ops4j/pax/cdi
./repository/org/ops4j/pax/cdi/pax-cdi-features
./repository/org/ops4j/pax/cdi/pax-cdi-features/0.12.0
```

The first seem to be the feature repositories that are referenced in my features.xml, and I assume the ops4j ones are dependencies. This points towards karaf-maven-plugin downloading artifacts into the wrong location during build. I've built both karaf/features and karaf/apache-brooklyn separately, and the offending behavior is generated by the karaf assembly project.

CMoH commented:
Seems to be fixed in 4.0.2: https://issues.apache.org/jira/browse/KARAF-4006

Testing (hopefully it won't take 2 hrs again to download all artifacts)

CMoH commented:
Using karaf-4.0.2 does fix the distro problem, but now pax-itest crashes (as predicted by @splatch).

@hzbarcea Ready for the magic incantation for pax-exam - if/when available :)

CMoH commented:
Please review #992 for solutions to the comments above.

bostko commented on pom.xml:
```diff
@@ -72,11 +72,122 @@
+        <maxmind.version>0.8.1</maxmind.version>
+        <jna.version>4.0.0</jna.version>
+        <winrm4j.version>0.1.0</winrm4j.version>
+        <coverage.target>${working.dir}</coverage.target>
```
@CMoH why is this property needed?

hzbarcea replied:
Good spot. Not needed. Also the version of antrun used is 1.8 in ./parent.pom.xml. Either needs to be removed, or changed to 1.8 and modify ./parent/pom.xml to use the version propert. Thanks @bostko .
